apps/server/lib/lexical/server/provider/handlers/commands.ex

```diff
@@ -0,0 +1,65 @@
+defmodule Lexical.Server.Provider.Handlers.Commands do
+  alias Lexical.Formats
+  alias Lexical.Project
+  alias Lexical.Protocol.Requests
+  alias Lexical.Protocol.Responses
+  alias Lexical.Protocol.Types
+  alias Lexical.Protocol.Types.ErrorCodes
+  alias Lexical.RemoteControl
+  alias Lexical.Server.Provider.Env
+  alias Lexical.Server.Window
+
+  require ErrorCodes
+  require Logger
+
+  def names do
+    [reindex_name()]
+  end
+
+  def reindex_command(%Project{} = project) do
+    project_name = Project.name(project)
+
+    Types.Command.new(
+      title: "Rebuild #{project_name}'s code search index",
+      command: reindex_name()
+    )
+  end
+
+  def handle(%Requests.ExecuteCommand{} = request, %Env{} = env) do
+    response =
+      case request.command do
+        "Reindex" ->
+          Logger.info("Reindex #{Project.name(env.project)}")
+          reindex(env.project, request.id)
+
+        invalid ->
+          message = "#{invalid} is not a valid command"
+          internal_error(request.id, message)
+      end
+
+    {:reply, response}
+  end
+
+  defp reindex(%Project{} = project, request_id) do
+    case :timer.tc(RemoteControl.Api, :reindex, [project]) do
+      {elapsed, :ok} ->
+        message = "Reindexed #{Project.name(project)} in #{Formats.time(elapsed)}"
+        Window.show(:info, message)
+        Responses.ExecuteCommand.new(request_id, "ok")
+
+      {_elapsed, error} ->
+        Window.error("Indexing #{Project.name(project)} failed")
+        Logger.error("Indexing command failed due to #{inspect(error)}")
+
+        internal_error(request_id, "Indexing Failed")
```

Comment:
I've noticed that when I click too quickly, it fails here. I believe that `{:error, "Already Running"}` should be set to info: `"Already running"` instead of the current error failed message. It would result in a better user experience.

![CleanShot 2023-12-16 at 15 49 31@2x](https://github.com/lexical-lsp/lexical/assets/12830256/7ce3a3d8-5435-4c24-983b-a9ee25b7b572)

Reply:
While I agree with this, I'd probably go even further and say that another reindex command should just cancel the current one and restart it.

Reply:
Why not let the first one finish?

Reply:
Cancelling the task will mean that the index could be in an inconsistent state. Letting the first task finish ensures that the index will always exist, and that it is correct.

Reply:
I was considering triggering a recompute of code lenses when indexing starts and stops and not emitting this lens if there's an index job running.

Reply:
> Cancelling the task will mean that the index could be in an inconsistent state. Letting the first task finish ensures that the index will always exist, and that it is correct.

Oh, I thought all entries were written together at the same time.

My assumption is that, since this is a manually-triggered command, if you're triggering it again while an index is in-process, you're doing it on purpose and the result of the in-process one isn't useful to you (e.g. you changed something and want it reflected in the index).

Reply:
> Oh, I thought all entries were written together at the same time.

That's true, but if you exit the process while the write is happening, you will lose data (this is surprisingly easy to do).

Reply:
> e.g. you changed something and want it reflected in the index

this is not a use case. Changing something updates the index automatically. The use case for this button is "I've upgraded lexical, and I need to reindex my source code to get some new feature". Mainly, this is for us.